File: labs/control-ui/api-shadow/hybrid_shadow_v2_transport.ts
export const HybridShadowCipher = {
  AeadAes256Gcm: 'AEAD_AES_256_GCM',
  AeadChacha20Poly1305: 'AEAD_CHACHA20_POLY1305',
} as const;
export type HybridShadowCipher = (typeof HybridShadowCipher)[keyof typeof HybridShadowCipher];

export interface HybridShadowConfig {
  cipher?: HybridShadowCipher;
  psk: Uint8Array;
  saltLength?: number;
  replayWindowSecs?: number;
}

export class HybridShadowV2Transport {
  public config: Required<HybridShadowConfig>;
  private saltHistory: Map<string, number> = new Map();

  constructor(config: HybridShadowConfig) {
    if (config.psk.length < 16) {
      throw new Error('PSK must be at least 16 bytes');
    }
    this.config = {
      cipher: config.cipher ?? HybridShadowCipher.AeadAes256Gcm,
      psk: config.psk,
      saltLength: config.saltLength ?? 32,
      replayWindowSecs: config.replayWindowSecs ?? 120,
    };
  }

  generateSalt(): Uint8Array {
    const salt = new Uint8Array(this.config.saltLength);
    if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
      crypto.getRandomValues(salt);
    } else {
      for (let i = 0; i < salt.length; i++) {
        salt[i] = Math.floor(Math.random() * 256);
      }
    }
    return salt;
  }

  deriveSubkey(salt: Uint8Array): Uint8Array {
    const subkey = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      const pskByte = this.config.psk[i % this.config.psk.length]!;
      const saltByte = salt[i % salt.length]!;
      subkey[i] = (pskByte ^ saltByte ^ (i * 7 + 13)) & 0xff;
    }
    return subkey;
  }

  registerSalt(salt: Uint8Array, nowSecs: number): boolean {
    const cutoff = nowSecs - this.config.replayWindowSecs;
    for (const [key, ts] of this.saltHistory.entries()) {
      if (ts < cutoff) {
        this.saltHistory.delete(key);
      }
    }

    const hex = Array.from(salt)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
    if (this.saltHistory.has(hex)) {
      return false; // Replay detected
    }

    this.saltHistory.set(hex, nowSecs);
    return true;
  }

  framePayload(salt: Uint8Array, payload: Uint8Array): Uint8Array {
    const frame = new Uint8Array(1 + salt.length + 2 + payload.length);
    frame[0] = salt.length;
    frame.set(salt, 1);
    const pLen = payload.length;
    frame[1 + salt.length] = (pLen >> 8) & 0xff;
    frame[2 + salt.length] = pLen & 0xff;
    frame.set(payload, 3 + salt.length);
    return frame;
  }

  unframePayload(data: Uint8Array): { salt: Uint8Array; payload: Uint8Array } | null {
    if (data.length < 3) return null;
    const saltLen = data[0]!;
    if (data.length < 1 + saltLen + 2) return null;

    const salt = data.slice(1, 1 + saltLen);
    const pLen = (data[1 + saltLen]! << 8) | data[2 + saltLen]!;
    const start = 3 + saltLen;
    if (data.length < start + pLen) return null;

    const payload = data.slice(start, start + pLen);
    return { salt, payload };
  }
}

File: labs/mobile/donor-archive/tunnel/HybridShadowV2Transport.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

enum class HybridShadowCipher {
    AEAD_AES_256_GCM,
    AEAD_CHACHA20_POLY1305
}

data class HybridShadowConfig(
    val cipher: HybridShadowCipher = HybridShadowCipher.AEAD_AES_256_GCM,
    val psk: ByteArray,
    val saltLength: Int = 32,
    val replayWindowSecs: Long = 120
)

class HybridShadowV2Transport(private val config: HybridShadowConfig) {
    private val saltHistory = mutableMapOf<String, Long>()
    private val random = SecureRandom()

    fun generateSalt(): ByteArray {
        val salt = ByteArray(config.saltLength)
        random.nextBytes(salt)
        return salt
    }

    fun deriveSubkey(salt: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(config.psk, "HmacSHA256"))
        mac.update(salt)
        return mac.doFinal("hybrid-shadow-v2-subkey".toByteArray())
    }

    @Synchronized
    fun registerSalt(salt: ByteArray, nowSecs: Long): Boolean {
        // Prune stale
        val cutoff = nowSecs - config.replayWindowSecs
        val iterator = saltHistory.entries.iterator()
        while (iterator.hasNext()) {
            if (iterator.next().value < cutoff) {
                iterator.remove()
            }
        }

        val key = salt.joinToString("") { "%02x".format(it) }
        if (saltHistory.containsKey(key)) {
            return false // Replay
        }
        saltHistory[key] = nowSecs
        return true
    }

    fun framePayload(salt: ByteArray, payload: ByteArray): ByteArray {
        val buf = ByteBuffer.allocate(1 + salt.size + 2 + payload.size)
        buf.put(salt.size.toByte())
        buf.put(salt)
        buf.putShort(payload.size.toShort())
        buf.put(payload)
        return buf.array()
    }

    fun unframePayload(data: ByteArray): Pair<ByteArray, ByteArray>? {
        if (data.size < 3) return null
        val saltLen = data[0].toInt() and 0xFF
        if (data.size < 1 + saltLen + 2) return null

        val salt = data.copyOfRange(1, 1 + saltLen)
        val pLen = ((data[1 + saltLen].toInt() and 0xFF) shl 8) or (data[2 + saltLen].toInt() and 0xFF)
        val start = 3 + saltLen
        if (data.size < start + pLen) return null

        val payload = data.copyOfRange(start, start + pLen)
        return Pair(salt, payload)
    }
}
